// @effect-ts/system/Layer/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  all: true,
  allSeq: true,
  and: true,
  andSeq: true,
  andSeq_: true,
  and_: true,
  build: true,
  catchAll: true,
  chain: true,
  chain_: true,
  compose: true,
  compose_: true,
  Empty: true,
  fail: true,
  first: true,
  flatten: true,
  fold: true,
  fresh: true,
  fromEffect: true,
  fromEffect_: true,
  fromFunction: true,
  fromManaged: true,
  fromManaged_: true,
  fromRawEffect: true,
  fromRawFunction: true,
  fromRawFunctionM: true,
  fromRawManaged: true,
  identity: true,
  launch: true,
  Layer: true,
  main: true,
  makeMemoMap: true,
  map: true,
  mapError: true,
  map_: true,
  MemoMap: true,
  orDie: true,
  orElse: true,
  fromValue: true,
  restrict: true,
  retry: true,
  scope: true,
  second: true,
  succeed: true,
  suspend: true,
  toRuntime: true,
  using: true,
  usingAnd: true,
  zip: true,
  zipPar: true,
  zipPar_: true,
  zipWithPar: true,
  zipWithPar_: true,
  zip_: true
};
Object.defineProperty(exports, "Empty", {
  enumerable: true,
  get: function () {
    return _core.Empty;
  }
});
Object.defineProperty(exports, "Layer", {
  enumerable: true,
  get: function () {
    return _core.Layer;
  }
});
Object.defineProperty(exports, "MemoMap", {
  enumerable: true,
  get: function () {
    return _core.MemoMap;
  }
});
Object.defineProperty(exports, "all", {
  enumerable: true,
  get: function () {
    return _core.all;
  }
});
Object.defineProperty(exports, "allSeq", {
  enumerable: true,
  get: function () {
    return _core.allSeq;
  }
});
Object.defineProperty(exports, "and", {
  enumerable: true,
  get: function () {
    return _core.and;
  }
});
Object.defineProperty(exports, "andSeq", {
  enumerable: true,
  get: function () {
    return _core.andSeq;
  }
});
Object.defineProperty(exports, "andSeq_", {
  enumerable: true,
  get: function () {
    return _core.andSeq_;
  }
});
Object.defineProperty(exports, "and_", {
  enumerable: true,
  get: function () {
    return _core.and_;
  }
});
Object.defineProperty(exports, "build", {
  enumerable: true,
  get: function () {
    return _core.build;
  }
});
Object.defineProperty(exports, "catchAll", {
  enumerable: true,
  get: function () {
    return _core.catchAll;
  }
});
Object.defineProperty(exports, "chain", {
  enumerable: true,
  get: function () {
    return _core.chain;
  }
});
Object.defineProperty(exports, "chain_", {
  enumerable: true,
  get: function () {
    return _core.chain_;
  }
});
Object.defineProperty(exports, "compose", {
  enumerable: true,
  get: function () {
    return _core.compose;
  }
});
Object.defineProperty(exports, "compose_", {
  enumerable: true,
  get: function () {
    return _core.compose_;
  }
});
Object.defineProperty(exports, "fail", {
  enumerable: true,
  get: function () {
    return _core.fail;
  }
});
Object.defineProperty(exports, "first", {
  enumerable: true,
  get: function () {
    return _core.first;
  }
});
Object.defineProperty(exports, "flatten", {
  enumerable: true,
  get: function () {
    return _core.flatten;
  }
});
Object.defineProperty(exports, "fold", {
  enumerable: true,
  get: function () {
    return _core.fold;
  }
});
Object.defineProperty(exports, "fresh", {
  enumerable: true,
  get: function () {
    return _core.fresh;
  }
});
Object.defineProperty(exports, "fromEffect", {
  enumerable: true,
  get: function () {
    return _core.fromEffect;
  }
});
Object.defineProperty(exports, "fromEffect_", {
  enumerable: true,
  get: function () {
    return _core.fromEffect_;
  }
});
Object.defineProperty(exports, "fromFunction", {
  enumerable: true,
  get: function () {
    return _core.fromFunction;
  }
});
Object.defineProperty(exports, "fromManaged", {
  enumerable: true,
  get: function () {
    return _core.fromManaged;
  }
});
Object.defineProperty(exports, "fromManaged_", {
  enumerable: true,
  get: function () {
    return _core.fromManaged_;
  }
});
Object.defineProperty(exports, "fromRawEffect", {
  enumerable: true,
  get: function () {
    return _core.fromRawEffect;
  }
});
Object.defineProperty(exports, "fromRawFunction", {
  enumerable: true,
  get: function () {
    return _core.fromRawFunction;
  }
});
Object.defineProperty(exports, "fromRawFunctionM", {
  enumerable: true,
  get: function () {
    return _core.fromRawFunctionM;
  }
});
Object.defineProperty(exports, "fromRawManaged", {
  enumerable: true,
  get: function () {
    return _core.fromRawManaged;
  }
});
Object.defineProperty(exports, "fromValue", {
  enumerable: true,
  get: function () {
    return _core.fromValue;
  }
});
Object.defineProperty(exports, "identity", {
  enumerable: true,
  get: function () {
    return _core.identity;
  }
});
Object.defineProperty(exports, "launch", {
  enumerable: true,
  get: function () {
    return _core.launch;
  }
});
Object.defineProperty(exports, "main", {
  enumerable: true,
  get: function () {
    return _core.main;
  }
});
Object.defineProperty(exports, "makeMemoMap", {
  enumerable: true,
  get: function () {
    return _core.makeMemoMap;
  }
});
Object.defineProperty(exports, "map", {
  enumerable: true,
  get: function () {
    return _core.map;
  }
});
Object.defineProperty(exports, "mapError", {
  enumerable: true,
  get: function () {
    return _core.mapError;
  }
});
Object.defineProperty(exports, "map_", {
  enumerable: true,
  get: function () {
    return _core.map_;
  }
});
Object.defineProperty(exports, "orDie", {
  enumerable: true,
  get: function () {
    return _core.orDie;
  }
});
Object.defineProperty(exports, "orElse", {
  enumerable: true,
  get: function () {
    return _core.orElse;
  }
});
Object.defineProperty(exports, "restrict", {
  enumerable: true,
  get: function () {
    return _core.restrict;
  }
});
Object.defineProperty(exports, "retry", {
  enumerable: true,
  get: function () {
    return _core.retry;
  }
});
Object.defineProperty(exports, "scope", {
  enumerable: true,
  get: function () {
    return _core.scope;
  }
});
Object.defineProperty(exports, "second", {
  enumerable: true,
  get: function () {
    return _core.second;
  }
});
Object.defineProperty(exports, "succeed", {
  enumerable: true,
  get: function () {
    return _core.succeed;
  }
});
Object.defineProperty(exports, "suspend", {
  enumerable: true,
  get: function () {
    return _core.suspend;
  }
});
Object.defineProperty(exports, "toRuntime", {
  enumerable: true,
  get: function () {
    return _core.toRuntime;
  }
});
Object.defineProperty(exports, "using", {
  enumerable: true,
  get: function () {
    return _core.using;
  }
});
Object.defineProperty(exports, "usingAnd", {
  enumerable: true,
  get: function () {
    return _core.usingAnd;
  }
});
Object.defineProperty(exports, "zip", {
  enumerable: true,
  get: function () {
    return _core.zip;
  }
});
Object.defineProperty(exports, "zipPar", {
  enumerable: true,
  get: function () {
    return _core.zipPar;
  }
});
Object.defineProperty(exports, "zipPar_", {
  enumerable: true,
  get: function () {
    return _core.zipPar_;
  }
});
Object.defineProperty(exports, "zipWithPar", {
  enumerable: true,
  get: function () {
    return _core.zipWithPar;
  }
});
Object.defineProperty(exports, "zipWithPar_", {
  enumerable: true,
  get: function () {
    return _core.zipWithPar_;
  }
});
Object.defineProperty(exports, "zip_", {
  enumerable: true,
  get: function () {
    return _core.zip_;
  }
});

require("../Operator/index.js");

var _core = /*#__PURE__*/require("./core.js");

var _unsafe = /*#__PURE__*/require("./unsafe.js");

Object.keys(_unsafe).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _unsafe[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _unsafe[key];
    }
  });
});
