// @effect-ts/core/_mjs/Modules/index.mjs
// ets_tracing: off
import "../Operator/index.mjs";
export { EffectURI } from "@effect-ts/system/Effect";
export const ArrayURI = "Array";
export const NonEmptyArrayURI = "NonEmptyArray";
export const BoundedURI = "Bounded";
export const ClosureURI = "Closure";
export const CommutativeURI = "Commutative";
export const EitherURI = "Either";
export const EqualURI = "Equal";
export const IdentityURI = "Identity";
export const InverseURI = "Inverse";
export const IterableURI = "IterableURI";
export const OrdURI = "Ord";
export const DictionaryURI = "Dictionary";
export const ShowURI = "Show";
export const XIOURI = "XIO";
export const XReaderURI = "XReader";
export const XStateURI = "XState";
export const ReaderURI = "Reader";
export const StateInURI = "StateIn";
export const ParametricStateInURI = "ParametricStateIn";
export const StateOutURI = "StateOut";
export const IxURI = "Ix";
export const TaskURI = "Task";
export const ConstURI = "Const";
export const LayerURI = "Layer";
export const SyncURI = "Sync";
export const XPureURI = "XPure";
export const IOURI = "IO";
export const XPureReaderCategoryURI = "XPureReaderCategory";
export const XPureStateCategoryURI = "XPureStateCategory";
export const IdURI = "Id";
export const EffectCategoryURI = "EffectCategory";
export const FreeAssociativeURI = "FreeAssociative";
export const AsyncURI = "Async";
export const TreeURI = "Tree";
export const OptionURI = "Option";
export const ListURI = "List";
export const ChunkURI = "Chunk";
