// @effect-ts/core/Modules/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EffectCategoryURI = exports.DictionaryURI = exports.ConstURI = exports.CommutativeURI = exports.ClosureURI = exports.ChunkURI = exports.BoundedURI = exports.AsyncURI = exports.ArrayURI = void 0;
Object.defineProperty(exports, "EffectURI", {
  enumerable: true,
  get: function () {
    return _Effect.EffectURI;
  }
});
exports.XStateURI = exports.XReaderURI = exports.XPureURI = exports.XPureStateCategoryURI = exports.XPureReaderCategoryURI = exports.XIOURI = exports.TreeURI = exports.TaskURI = exports.SyncURI = exports.StateOutURI = exports.StateInURI = exports.ShowURI = exports.ReaderURI = exports.ParametricStateInURI = exports.OrdURI = exports.OptionURI = exports.NonEmptyArrayURI = exports.ListURI = exports.LayerURI = exports.IxURI = exports.IterableURI = exports.InverseURI = exports.IdentityURI = exports.IdURI = exports.IOURI = exports.FreeAssociativeURI = exports.EqualURI = exports.EitherURI = void 0;

require("../Operator/index.js");

var _Effect = /*#__PURE__*/require("@effect-ts/system/Effect");

// ets_tracing: off
const ArrayURI = "Array";
exports.ArrayURI = ArrayURI;
const NonEmptyArrayURI = "NonEmptyArray";
exports.NonEmptyArrayURI = NonEmptyArrayURI;
const BoundedURI = "Bounded";
exports.BoundedURI = BoundedURI;
const ClosureURI = "Closure";
exports.ClosureURI = ClosureURI;
const CommutativeURI = "Commutative";
exports.CommutativeURI = CommutativeURI;
const EitherURI = "Either";
exports.EitherURI = EitherURI;
const EqualURI = "Equal";
exports.EqualURI = EqualURI;
const IdentityURI = "Identity";
exports.IdentityURI = IdentityURI;
const InverseURI = "Inverse";
exports.InverseURI = InverseURI;
const IterableURI = "IterableURI";
exports.IterableURI = IterableURI;
const OrdURI = "Ord";
exports.OrdURI = OrdURI;
const DictionaryURI = "Dictionary";
exports.DictionaryURI = DictionaryURI;
const ShowURI = "Show";
exports.ShowURI = ShowURI;
const XIOURI = "XIO";
exports.XIOURI = XIOURI;
const XReaderURI = "XReader";
exports.XReaderURI = XReaderURI;
const XStateURI = "XState";
exports.XStateURI = XStateURI;
const ReaderURI = "Reader";
exports.ReaderURI = ReaderURI;
const StateInURI = "StateIn";
exports.StateInURI = StateInURI;
const ParametricStateInURI = "ParametricStateIn";
exports.ParametricStateInURI = ParametricStateInURI;
const StateOutURI = "StateOut";
exports.StateOutURI = StateOutURI;
const IxURI = "Ix";
exports.IxURI = IxURI;
const TaskURI = "Task";
exports.TaskURI = TaskURI;
const ConstURI = "Const";
exports.ConstURI = ConstURI;
const LayerURI = "Layer";
exports.LayerURI = LayerURI;
const SyncURI = "Sync";
exports.SyncURI = SyncURI;
const XPureURI = "XPure";
exports.XPureURI = XPureURI;
const IOURI = "IO";
exports.IOURI = IOURI;
const XPureReaderCategoryURI = "XPureReaderCategory";
exports.XPureReaderCategoryURI = XPureReaderCategoryURI;
const XPureStateCategoryURI = "XPureStateCategory";
exports.XPureStateCategoryURI = XPureStateCategoryURI;
const IdURI = "Id";
exports.IdURI = IdURI;
const EffectCategoryURI = "EffectCategory";
exports.EffectCategoryURI = EffectCategoryURI;
const FreeAssociativeURI = "FreeAssociative";
exports.FreeAssociativeURI = FreeAssociativeURI;
const AsyncURI = "Async";
exports.AsyncURI = AsyncURI;
const TreeURI = "Tree";
exports.TreeURI = TreeURI;
const OptionURI = "Option";
exports.OptionURI = OptionURI;
const ListURI = "List";
exports.ListURI = ListURI;
const ChunkURI = "Chunk";
exports.ChunkURI = ChunkURI;
