// @contentlayer/cli/dist/commands/_BaseCommand.js
import * as core from '@contentlayer/core';
import { fs } from '@contentlayer/utils';
import { pipe, T } from '@contentlayer/utils/effect';
import { Command, Option } from 'clipanion';
import * as t from 'typanion';
export class BaseCommand extends Command {
    constructor() {
        super(...arguments);
        this.configPath = Option.String('-c,--config', {
            description: 'Path to the config (default: contentlayer.config.ts/js)',
            validator: t.isString(),
            required: false,
        });
        this.clearCache = Option.Boolean('--clearCache', false, {
            description: 'Whether to clear the cache before running the command',
        });
        this.verbose = Option.Boolean('--verbose', false, {
            description: 'More verbose logging and error stack traces',
        });
        this.execute = () => pipe(this.executeSafe(), core.runMain({
            tracingServiceName: 'contentlayer-cli',
            verbose: this.verbose || process.env.CL_DEBUG !== undefined,
        }));
        this.clearCacheIfNeeded = () => {
            const { clearCache } = this;
            return T.gen(function* ($) {
                if (clearCache) {
                    const cwd = yield* $(core.getCwd);
                    const artifactsDir = core.ArtifactsDir.getDirPath({ cwd });
                    yield* $(fs.rm(artifactsDir, { recursive: true, force: true }));
                    yield* $(T.log('Cache cleared successfully'));
                }
            });
        };
    }
}
