// @effect-ts/system/Effect/excl-forEach-promise.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Promise", {
  enumerable: true,
  get: function () {
    return _promise.Promise;
  }
});
Object.defineProperty(exports, "await", {
  enumerable: true,
  get: function () {
    return _await.await;
  }
});
Object.defineProperty(exports, "fail", {
  enumerable: true,
  get: function () {
    return _fail.fail;
  }
});
Object.defineProperty(exports, "halt", {
  enumerable: true,
  get: function () {
    return _halt.halt;
  }
});
Object.defineProperty(exports, "interruptAs", {
  enumerable: true,
  get: function () {
    return _interruptAs.interruptAs;
  }
});
Object.defineProperty(exports, "make", {
  enumerable: true,
  get: function () {
    return _make.make;
  }
});
Object.defineProperty(exports, "succeed", {
  enumerable: true,
  get: function () {
    return _succeed.succeed;
  }
});
Object.defineProperty(exports, "unsafeDone", {
  enumerable: true,
  get: function () {
    return _unsafeDone.unsafeDone;
  }
});
Object.defineProperty(exports, "unsafeMake", {
  enumerable: true,
  get: function () {
    return _unsafeMake.unsafeMake;
  }
});

var _make = /*#__PURE__*/require("../Promise/make.js");

var _fail = /*#__PURE__*/require("../Promise/fail.js");

var _succeed = /*#__PURE__*/require("../Promise/succeed.js");

var _await = /*#__PURE__*/require("../Promise/await.js");

var _promise = /*#__PURE__*/require("../Promise/promise.js");

var _halt = /*#__PURE__*/require("../Promise/halt.js");

var _unsafeDone = /*#__PURE__*/require("../Promise/unsafeDone.js");

var _unsafeMake = /*#__PURE__*/require("../Promise/unsafeMake.js");

var _interruptAs = /*#__PURE__*/require("../Promise/interruptAs.js");
