// @effect-ts/system/Effect/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  collectAll: true,
  collectAllPar: true,
  collectAllParN: true,
  collectAllParN_: true,
  collectAllSuccesses: true,
  collectAllSuccessesPar: true,
  collectAllSuccessesParN: true,
  collectAllUnit: true,
  collectAllUnitPar: true,
  collectAllUnitParN: true,
  collectAllUnitParN_: true,
  collectAllWith: true,
  collectAllWithPar: true,
  collectAllWithParN: true,
  collectAllWithParN_: true,
  collectAllWithPar_: true,
  collectAllWith_: true,
  forEach: true,
  forEachWithIndex: true,
  forEachExec: true,
  forEachExec_: true,
  forEachPar: true,
  forEachParWithIndex: true,
  forEachParN_: true,
  forEachParWithIndexN_: true,
  forEachParN: true,
  forEachParWithIndexN: true,
  forEachPar_: true,
  forEachParWithIndex_: true,
  forEachUnit: true,
  forEachUnitPar: true,
  forEachUnitParN: true,
  forEachUnitParN_: true,
  forEachUnitPar_: true,
  forEachUnit_: true,
  forEach_: true,
  forEachWithIndex_: true,
  forkManaged: true
};
Object.defineProperty(exports, "collectAll", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAll;
  }
});
Object.defineProperty(exports, "collectAllPar", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllPar;
  }
});
Object.defineProperty(exports, "collectAllParN", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllParN;
  }
});
Object.defineProperty(exports, "collectAllParN_", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllParN_;
  }
});
Object.defineProperty(exports, "collectAllSuccesses", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllSuccesses;
  }
});
Object.defineProperty(exports, "collectAllSuccessesPar", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllSuccessesPar;
  }
});
Object.defineProperty(exports, "collectAllSuccessesParN", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllSuccessesParN;
  }
});
Object.defineProperty(exports, "collectAllUnit", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllUnit;
  }
});
Object.defineProperty(exports, "collectAllUnitPar", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllUnitPar;
  }
});
Object.defineProperty(exports, "collectAllUnitParN", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllUnitParN;
  }
});
Object.defineProperty(exports, "collectAllUnitParN_", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllUnitParN_;
  }
});
Object.defineProperty(exports, "collectAllWith", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllWith;
  }
});
Object.defineProperty(exports, "collectAllWithPar", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllWithPar;
  }
});
Object.defineProperty(exports, "collectAllWithParN", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllWithParN;
  }
});
Object.defineProperty(exports, "collectAllWithParN_", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllWithParN_;
  }
});
Object.defineProperty(exports, "collectAllWithPar_", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllWithPar_;
  }
});
Object.defineProperty(exports, "collectAllWith_", {
  enumerable: true,
  get: function () {
    return _exclForEach.collectAllWith_;
  }
});
Object.defineProperty(exports, "forEach", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEach;
  }
});
Object.defineProperty(exports, "forEachExec", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachExec;
  }
});
Object.defineProperty(exports, "forEachExec_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachExec_;
  }
});
Object.defineProperty(exports, "forEachPar", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachPar;
  }
});
Object.defineProperty(exports, "forEachParN", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachParN;
  }
});
Object.defineProperty(exports, "forEachParN_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachParN_;
  }
});
Object.defineProperty(exports, "forEachParWithIndex", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachParWithIndex;
  }
});
Object.defineProperty(exports, "forEachParWithIndexN", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachParWithIndexN;
  }
});
Object.defineProperty(exports, "forEachParWithIndexN_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachParWithIndexN_;
  }
});
Object.defineProperty(exports, "forEachParWithIndex_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachParWithIndex_;
  }
});
Object.defineProperty(exports, "forEachPar_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachPar_;
  }
});
Object.defineProperty(exports, "forEachUnit", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachUnit;
  }
});
Object.defineProperty(exports, "forEachUnitPar", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachUnitPar;
  }
});
Object.defineProperty(exports, "forEachUnitParN", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachUnitParN;
  }
});
Object.defineProperty(exports, "forEachUnitParN_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachUnitParN_;
  }
});
Object.defineProperty(exports, "forEachUnitPar_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachUnitPar_;
  }
});
Object.defineProperty(exports, "forEachUnit_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachUnit_;
  }
});
Object.defineProperty(exports, "forEachWithIndex", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachWithIndex;
  }
});
Object.defineProperty(exports, "forEachWithIndex_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEachWithIndex_;
  }
});
Object.defineProperty(exports, "forEach_", {
  enumerable: true,
  get: function () {
    return _exclForEach.forEach_;
  }
});
Object.defineProperty(exports, "forkManaged", {
  enumerable: true,
  get: function () {
    return _exclForEach.forkManaged;
  }
});

require("../Operator/index.js");

var _absolve = /*#__PURE__*/require("./absolve.js");

Object.keys(_absolve).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _absolve[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _absolve[key];
    }
  });
});

var _absorbWith = /*#__PURE__*/require("./absorbWith.js");

Object.keys(_absorbWith).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _absorbWith[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _absorbWith[key];
    }
  });
});

var _andThen = /*#__PURE__*/require("./andThen.js");

Object.keys(_andThen).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _andThen[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _andThen[key];
    }
  });
});

var _ap = /*#__PURE__*/require("./ap.js");

Object.keys(_ap).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ap[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ap[key];
    }
  });
});

var _as = /*#__PURE__*/require("./as.js");

Object.keys(_as).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _as[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _as[key];
    }
  });
});

var _asService = /*#__PURE__*/require("./asService.js");

Object.keys(_asService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _asService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _asService[key];
    }
  });
});

var _asSome = /*#__PURE__*/require("./asSome.js");

Object.keys(_asSome).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _asSome[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _asSome[key];
    }
  });
});

var _asSomeError = /*#__PURE__*/require("./asSomeError.js");

Object.keys(_asSomeError).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _asSomeError[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _asSomeError[key];
    }
  });
});

var _asUnit = /*#__PURE__*/require("./asUnit.js");

Object.keys(_asUnit).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _asUnit[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _asUnit[key];
    }
  });
});

var _awaitAllChildren = /*#__PURE__*/require("./awaitAllChildren.js");

Object.keys(_awaitAllChildren).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _awaitAllChildren[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _awaitAllChildren[key];
    }
  });
});

var _bimap = /*#__PURE__*/require("./bimap.js");

Object.keys(_bimap).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bimap[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bimap[key];
    }
  });
});

var _bindAll = /*#__PURE__*/require("./bindAll.js");

Object.keys(_bindAll).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bindAll[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bindAll[key];
    }
  });
});

var _bracket = /*#__PURE__*/require("./bracket.js");

Object.keys(_bracket).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bracket[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bracket[key];
    }
  });
});

var _bracketExit = /*#__PURE__*/require("./bracketExit.js");

Object.keys(_bracketExit).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bracketExit[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bracketExit[key];
    }
  });
});

var _bracketFiber = /*#__PURE__*/require("./bracketFiber.js");

Object.keys(_bracketFiber).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bracketFiber[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bracketFiber[key];
    }
  });
});

var _bracketOnError = /*#__PURE__*/require("./bracketOnError.js");

Object.keys(_bracketOnError).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bracketOnError[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bracketOnError[key];
    }
  });
});

var _cached = /*#__PURE__*/require("./cached.js");

Object.keys(_cached).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _cached[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _cached[key];
    }
  });
});

var _cachedInvalidate = /*#__PURE__*/require("./cachedInvalidate.js");

Object.keys(_cachedInvalidate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _cachedInvalidate[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _cachedInvalidate[key];
    }
  });
});

var _Canceler = /*#__PURE__*/require("./Canceler.js");

Object.keys(_Canceler).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _Canceler[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _Canceler[key];
    }
  });
});

var _catchAll = /*#__PURE__*/require("./catchAll.js");

Object.keys(_catchAll).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _catchAll[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _catchAll[key];
    }
  });
});

var _catchAllCause = /*#__PURE__*/require("./catchAllCause.js");

Object.keys(_catchAllCause).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _catchAllCause[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _catchAllCause[key];
    }
  });
});

var _catchAllDefect = /*#__PURE__*/require("./catchAllDefect.js");

Object.keys(_catchAllDefect).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _catchAllDefect[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _catchAllDefect[key];
    }
  });
});

var _catchSome = /*#__PURE__*/require("./catchSome.js");

Object.keys(_catchSome).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _catchSome[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _catchSome[key];
    }
  });
});

var _catchSomeCause_ = /*#__PURE__*/require("./catchSomeCause_.js");

Object.keys(_catchSomeCause_).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _catchSomeCause_[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _catchSomeCause_[key];
    }
  });
});

var _catchSomeDefect = /*#__PURE__*/require("./catchSomeDefect.js");

Object.keys(_catchSomeDefect).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _catchSomeDefect[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _catchSomeDefect[key];
    }
  });
});

var _cause = /*#__PURE__*/require("./cause.js");

Object.keys(_cause).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _cause[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _cause[key];
    }
  });
});

var _Cb = /*#__PURE__*/require("./Cb.js");

Object.keys(_Cb).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _Cb[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _Cb[key];
    }
  });
});

var _chainError = /*#__PURE__*/require("./chainError.js");

Object.keys(_chainError).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _chainError[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _chainError[key];
    }
  });
});

var _collect = /*#__PURE__*/require("./collect.js");

Object.keys(_collect).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _collect[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _collect[key];
    }
  });
});

var _commons = /*#__PURE__*/require("./commons.js");

Object.keys(_commons).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _commons[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _commons[key];
    }
  });
});

var _compose = /*#__PURE__*/require("./compose.js");

Object.keys(_compose).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _compose[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _compose[key];
    }
  });
});

var _cond = /*#__PURE__*/require("./cond.js");

Object.keys(_cond).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _cond[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _cond[key];
    }
  });
});

var _continueOrFail = /*#__PURE__*/require("./continueOrFail.js");

Object.keys(_continueOrFail).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _continueOrFail[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _continueOrFail[key];
    }
  });
});

var _coreScope = /*#__PURE__*/require("./core-scope.js");

Object.keys(_coreScope).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _coreScope[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _coreScope[key];
    }
  });
});

var _core = /*#__PURE__*/require("./core.js");

Object.keys(_core).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _core[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _core[key];
    }
  });
});

var _delay = /*#__PURE__*/require("./delay.js");

Object.keys(_delay).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _delay[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _delay[key];
    }
  });
});

var _derive = /*#__PURE__*/require("./derive.js");

Object.keys(_derive).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _derive[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _derive[key];
    }
  });
});

var _descriptor = /*#__PURE__*/require("./descriptor.js");

Object.keys(_descriptor).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _descriptor[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _descriptor[key];
    }
  });
});

var _die = /*#__PURE__*/require("./die.js");

Object.keys(_die).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _die[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _die[key];
    }
  });
});

var _dieMessage = /*#__PURE__*/require("./dieMessage.js");

Object.keys(_dieMessage).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dieMessage[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _dieMessage[key];
    }
  });
});

var _do = /*#__PURE__*/require("./do.js");

Object.keys(_do).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _do[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _do[key];
    }
  });
});

var _done = /*#__PURE__*/require("./done.js");

Object.keys(_done).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _done[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _done[key];
    }
  });
});

var _dropWhile = /*#__PURE__*/require("./dropWhile.js");

Object.keys(_dropWhile).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dropWhile[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _dropWhile[key];
    }
  });
});

var _effect = /*#__PURE__*/require("./effect.js");

Object.keys(_effect).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _effect[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _effect[key];
    }
  });
});

var _effectAsync = /*#__PURE__*/require("./effectAsync.js");

Object.keys(_effectAsync).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _effectAsync[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _effectAsync[key];
    }
  });
});

var _effectAsyncInterrupt = /*#__PURE__*/require("./effectAsyncInterrupt.js");

Object.keys(_effectAsyncInterrupt).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _effectAsyncInterrupt[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _effectAsyncInterrupt[key];
    }
  });
});

var _effectAsyncM = /*#__PURE__*/require("./effectAsyncM.js");

Object.keys(_effectAsyncM).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _effectAsyncM[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _effectAsyncM[key];
    }
  });
});

var _effectMaybeAsyncInterrupt = /*#__PURE__*/require("./effectMaybeAsyncInterrupt.js");

Object.keys(_effectMaybeAsyncInterrupt).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _effectMaybeAsyncInterrupt[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _effectMaybeAsyncInterrupt[key];
    }
  });
});

var _either = /*#__PURE__*/require("./either.js");

Object.keys(_either).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _either[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _either[key];
    }
  });
});

var _ensuring = /*#__PURE__*/require("./ensuring.js");

Object.keys(_ensuring).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ensuring[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ensuring[key];
    }
  });
});

var _ensuringChildren = /*#__PURE__*/require("./ensuringChildren.js");

Object.keys(_ensuringChildren).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ensuringChildren[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ensuringChildren[key];
    }
  });
});

var _environment = /*#__PURE__*/require("./environment.js");

Object.keys(_environment).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _environment[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _environment[key];
    }
  });
});

var _eventually = /*#__PURE__*/require("./eventually.js");

Object.keys(_eventually).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _eventually[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _eventually[key];
    }
  });
});

var _ExecutionStrategy = /*#__PURE__*/require("./ExecutionStrategy.js");

Object.keys(_ExecutionStrategy).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ExecutionStrategy[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ExecutionStrategy[key];
    }
  });
});

var _fail = /*#__PURE__*/require("./fail.js");

Object.keys(_fail).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fail[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fail[key];
    }
  });
});

var _fiberId = /*#__PURE__*/require("./fiberId.js");

Object.keys(_fiberId).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fiberId[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fiberId[key];
    }
  });
});

var _filter = /*#__PURE__*/require("./filter.js");

Object.keys(_filter).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _filter[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _filter[key];
    }
  });
});

var _filterOr = /*#__PURE__*/require("./filterOr.js");

Object.keys(_filterOr).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _filterOr[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _filterOr[key];
    }
  });
});

var _first = /*#__PURE__*/require("./first.js");

Object.keys(_first).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _first[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _first[key];
    }
  });
});

var _firstSuccessOf = /*#__PURE__*/require("./firstSuccessOf.js");

Object.keys(_firstSuccessOf).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _firstSuccessOf[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _firstSuccessOf[key];
    }
  });
});

var _flatten = /*#__PURE__*/require("./flatten.js");

Object.keys(_flatten).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _flatten[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _flatten[key];
    }
  });
});

var _flattenErrorOption = /*#__PURE__*/require("./flattenErrorOption.js");

Object.keys(_flattenErrorOption).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _flattenErrorOption[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _flattenErrorOption[key];
    }
  });
});

var _flip = /*#__PURE__*/require("./flip.js");

Object.keys(_flip).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _flip[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _flip[key];
    }
  });
});

var _flipWith = /*#__PURE__*/require("./flipWith.js");

Object.keys(_flipWith).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _flipWith[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _flipWith[key];
    }
  });
});

var _fold = /*#__PURE__*/require("./fold.js");

Object.keys(_fold).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fold[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fold[key];
    }
  });
});

var _foldCause = /*#__PURE__*/require("./foldCause.js");

Object.keys(_foldCause).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _foldCause[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _foldCause[key];
    }
  });
});

var _foldM = /*#__PURE__*/require("./foldM.js");

Object.keys(_foldM).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _foldM[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _foldM[key];
    }
  });
});

var _forever = /*#__PURE__*/require("./forever.js");

Object.keys(_forever).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _forever[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _forever[key];
    }
  });
});

var _forkAll = /*#__PURE__*/require("./forkAll.js");

Object.keys(_forkAll).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _forkAll[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _forkAll[key];
    }
  });
});

var _forkAs = /*#__PURE__*/require("./forkAs.js");

Object.keys(_forkAs).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _forkAs[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _forkAs[key];
    }
  });
});

var _forkManaged = /*#__PURE__*/require("./forkManaged.js");

Object.keys(_forkManaged).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _forkManaged[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _forkManaged[key];
    }
  });
});

var _forkWithErrorHandler = /*#__PURE__*/require("./forkWithErrorHandler.js");

Object.keys(_forkWithErrorHandler).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _forkWithErrorHandler[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _forkWithErrorHandler[key];
    }
  });
});

var _fromEither = /*#__PURE__*/require("./fromEither.js");

Object.keys(_fromEither).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fromEither[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fromEither[key];
    }
  });
});

var _fromFiber = /*#__PURE__*/require("./fromFiber.js");

Object.keys(_fromFiber).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fromFiber[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fromFiber[key];
    }
  });
});

var _fromNodeCb = /*#__PURE__*/require("./fromNodeCb.js");

Object.keys(_fromNodeCb).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fromNodeCb[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fromNodeCb[key];
    }
  });
});

var _fromOption = /*#__PURE__*/require("./fromOption.js");

Object.keys(_fromOption).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fromOption[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fromOption[key];
    }
  });
});

var _fromPredicate = /*#__PURE__*/require("./fromPredicate.js");

Object.keys(_fromPredicate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _fromPredicate[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _fromPredicate[key];
    }
  });
});

var _gen = /*#__PURE__*/require("./gen.js");

Object.keys(_gen).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _gen[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _gen[key];
    }
  });
});

var _get = /*#__PURE__*/require("./get.js");

Object.keys(_get).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _get[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _get[key];
    }
  });
});

var _getOrFail = /*#__PURE__*/require("./getOrFail.js");

Object.keys(_getOrFail).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getOrFail[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getOrFail[key];
    }
  });
});

var _has = /*#__PURE__*/require("./has.js");

Object.keys(_has).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _has[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _has[key];
    }
  });
});

var _head = /*#__PURE__*/require("./head.js");

Object.keys(_head).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _head[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _head[key];
    }
  });
});

var _ifM = /*#__PURE__*/require("./ifM.js");

Object.keys(_ifM).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ifM[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ifM[key];
    }
  });
});

var _ignore = /*#__PURE__*/require("./ignore.js");

Object.keys(_ignore).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ignore[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ignore[key];
    }
  });
});

var _in = /*#__PURE__*/require("./in.js");

Object.keys(_in).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _in[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _in[key];
    }
  });
});

var _interop = /*#__PURE__*/require("./interop.js");

Object.keys(_interop).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interop[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interop[key];
    }
  });
});

var _interruptAllChildren = /*#__PURE__*/require("./interruptAllChildren.js");

Object.keys(_interruptAllChildren).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interruptAllChildren[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interruptAllChildren[key];
    }
  });
});

var _interruption = /*#__PURE__*/require("./interruption.js");

Object.keys(_interruption).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interruption[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interruption[key];
    }
  });
});

var _isFailure = /*#__PURE__*/require("./isFailure.js");

Object.keys(_isFailure).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _isFailure[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _isFailure[key];
    }
  });
});

var _isSuccess = /*#__PURE__*/require("./isSuccess.js");

Object.keys(_isSuccess).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _isSuccess[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _isSuccess[key];
    }
  });
});

var _iterate = /*#__PURE__*/require("./iterate.js");

Object.keys(_iterate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _iterate[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _iterate[key];
    }
  });
});

var _join = /*#__PURE__*/require("./join.js");

Object.keys(_join).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _join[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _join[key];
    }
  });
});

var _leftOrFail = /*#__PURE__*/require("./leftOrFail.js");

Object.keys(_leftOrFail).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _leftOrFail[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _leftOrFail[key];
    }
  });
});

var _loop = /*#__PURE__*/require("./loop.js");

Object.keys(_loop).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _loop[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _loop[key];
    }
  });
});

var _map = /*#__PURE__*/require("./map.js");

Object.keys(_map).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _map[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _map[key];
    }
  });
});

var _mapBoth = /*#__PURE__*/require("./mapBoth.js");

Object.keys(_mapBoth).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mapBoth[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mapBoth[key];
    }
  });
});

var _mapError = /*#__PURE__*/require("./mapError.js");

Object.keys(_mapError).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mapError[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mapError[key];
    }
  });
});

var _mapErrorCause = /*#__PURE__*/require("./mapErrorCause.js");

Object.keys(_mapErrorCause).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mapErrorCause[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mapErrorCause[key];
    }
  });
});

var _mapN = /*#__PURE__*/require("./mapN.js");

Object.keys(_mapN).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mapN[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mapN[key];
    }
  });
});

var _mapTryCatch = /*#__PURE__*/require("./mapTryCatch.js");

Object.keys(_mapTryCatch).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mapTryCatch[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mapTryCatch[key];
    }
  });
});

var _memoize = /*#__PURE__*/require("./memoize.js");

Object.keys(_memoize).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _memoize[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _memoize[key];
    }
  });
});

var _merge = /*#__PURE__*/require("./merge.js");

Object.keys(_merge).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _merge[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _merge[key];
    }
  });
});

var _mergeAll = /*#__PURE__*/require("./mergeAll.js");

Object.keys(_mergeAll).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mergeAll[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mergeAll[key];
    }
  });
});

var _never = /*#__PURE__*/require("./never.js");

Object.keys(_never).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _never[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _never[key];
    }
  });
});

var _none = /*#__PURE__*/require("./none.js");

Object.keys(_none).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _none[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _none[key];
    }
  });
});

var _once = /*#__PURE__*/require("./once.js");

Object.keys(_once).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _once[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _once[key];
    }
  });
});

var _onExit = /*#__PURE__*/require("./onExit.js");

Object.keys(_onExit).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _onExit[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _onExit[key];
    }
  });
});

var _onFirst = /*#__PURE__*/require("./onFirst.js");

Object.keys(_onFirst).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _onFirst[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _onFirst[key];
    }
  });
});

var _onLeft = /*#__PURE__*/require("./onLeft.js");

Object.keys(_onLeft).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _onLeft[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _onLeft[key];
    }
  });
});

var _onRight = /*#__PURE__*/require("./onRight.js");

Object.keys(_onRight).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _onRight[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _onRight[key];
    }
  });
});

var _onSecond = /*#__PURE__*/require("./onSecond.js");

Object.keys(_onSecond).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _onSecond[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _onSecond[key];
    }
  });
});

var _onTermination = /*#__PURE__*/require("./onTermination.js");

Object.keys(_onTermination).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _onTermination[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _onTermination[key];
    }
  });
});

var _option = /*#__PURE__*/require("./option.js");

Object.keys(_option).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _option[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _option[key];
    }
  });
});

var _optional = /*#__PURE__*/require("./optional.js");

Object.keys(_optional).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _optional[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _optional[key];
    }
  });
});

var _orDie = /*#__PURE__*/require("./orDie.js");

Object.keys(_orDie).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _orDie[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _orDie[key];
    }
  });
});

var _orDieKeep = /*#__PURE__*/require("./orDieKeep.js");

Object.keys(_orDieKeep).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _orDieKeep[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _orDieKeep[key];
    }
  });
});

var _orDieWith = /*#__PURE__*/require("./orDieWith.js");

Object.keys(_orDieWith).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _orDieWith[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _orDieWith[key];
    }
  });
});

var _orElse = /*#__PURE__*/require("./orElse.js");

Object.keys(_orElse).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _orElse[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _orElse[key];
    }
  });
});

var _orElseEither = /*#__PURE__*/require("./orElseEither.js");

Object.keys(_orElseEither).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _orElseEither[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _orElseEither[key];
    }
  });
});

var _orElseFail = /*#__PURE__*/require("./orElseFail.js");

Object.keys(_orElseFail).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _orElseFail[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _orElseFail[key];
    }
  });
});

var _orElseOptional = /*#__PURE__*/require("./orElseOptional.js");

Object.keys(_orElseOptional).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _orElseOptional[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _orElseOptional[key];
    }
  });
});

var _orElseSucceed = /*#__PURE__*/require("./orElseSucceed.js");

Object.keys(_orElseSucceed).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _orElseSucceed[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _orElseSucceed[key];
    }
  });
});

var _parallelErrors = /*#__PURE__*/require("./parallelErrors.js");

Object.keys(_parallelErrors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _parallelErrors[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _parallelErrors[key];
    }
  });
});

var _partition = /*#__PURE__*/require("./partition.js");

Object.keys(_partition).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _partition[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _partition[key];
    }
  });
});

var _primitives = /*#__PURE__*/require("./primitives.js");

Object.keys(_primitives).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _primitives[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _primitives[key];
    }
  });
});

var _promise = /*#__PURE__*/require("./promise.js");

Object.keys(_promise).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _promise[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _promise[key];
    }
  });
});

var _provide = /*#__PURE__*/require("./provide.js");

Object.keys(_provide).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _provide[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _provide[key];
    }
  });
});

var _provideManaged = /*#__PURE__*/require("./provideManaged.js");

Object.keys(_provideManaged).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _provideManaged[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _provideManaged[key];
    }
  });
});

var _provideSome = /*#__PURE__*/require("./provideSome.js");

Object.keys(_provideSome).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _provideSome[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _provideSome[key];
    }
  });
});

var _provideSomeLayer = /*#__PURE__*/require("./provideSomeLayer.js");

Object.keys(_provideSomeLayer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _provideSomeLayer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _provideSomeLayer[key];
    }
  });
});

var _race = /*#__PURE__*/require("./race.js");

Object.keys(_race).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _race[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _race[key];
    }
  });
});

var _raceAll = /*#__PURE__*/require("./raceAll.js");

Object.keys(_raceAll).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _raceAll[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _raceAll[key];
    }
  });
});

var _reduce = /*#__PURE__*/require("./reduce.js");

Object.keys(_reduce).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reduce[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _reduce[key];
    }
  });
});

var _reduceAll = /*#__PURE__*/require("./reduceAll.js");

Object.keys(_reduceAll).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reduceAll[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _reduceAll[key];
    }
  });
});

var _refailWithTrace = /*#__PURE__*/require("./refailWithTrace.js");

Object.keys(_refailWithTrace).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _refailWithTrace[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _refailWithTrace[key];
    }
  });
});

var _refineOrDie = /*#__PURE__*/require("./refineOrDie.js");

Object.keys(_refineOrDie).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _refineOrDie[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _refineOrDie[key];
    }
  });
});

var _reject = /*#__PURE__*/require("./reject.js");

Object.keys(_reject).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reject[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _reject[key];
    }
  });
});

var _repeat = /*#__PURE__*/require("./repeat.js");

Object.keys(_repeat).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _repeat[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _repeat[key];
    }
  });
});

var _repeatN = /*#__PURE__*/require("./repeatN.js");

Object.keys(_repeatN).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _repeatN[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _repeatN[key];
    }
  });
});

var _repeatUntil = /*#__PURE__*/require("./repeatUntil.js");

Object.keys(_repeatUntil).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _repeatUntil[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _repeatUntil[key];
    }
  });
});

var _repeatWhile = /*#__PURE__*/require("./repeatWhile.js");

Object.keys(_repeatWhile).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _repeatWhile[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _repeatWhile[key];
    }
  });
});

var _replicate = /*#__PURE__*/require("./replicate.js");

Object.keys(_replicate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _replicate[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _replicate[key];
    }
  });
});

var _require = /*#__PURE__*/require("./require.js");

Object.keys(_require).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _require[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _require[key];
    }
  });
});

var _reserve = /*#__PURE__*/require("./reserve.js");

Object.keys(_reserve).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reserve[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _reserve[key];
    }
  });
});

var _restrictions = /*#__PURE__*/require("./restrictions.js");

Object.keys(_restrictions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _restrictions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _restrictions[key];
    }
  });
});

var _resurrect = /*#__PURE__*/require("./resurrect.js");

Object.keys(_resurrect).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _resurrect[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _resurrect[key];
    }
  });
});

var _retry = /*#__PURE__*/require("./retry.js");

Object.keys(_retry).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _retry[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _retry[key];
    }
  });
});

var _retryUntil = /*#__PURE__*/require("./retryUntil.js");

Object.keys(_retryUntil).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _retryUntil[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _retryUntil[key];
    }
  });
});

var _retryWhile = /*#__PURE__*/require("./retryWhile.js");

Object.keys(_retryWhile).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _retryWhile[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _retryWhile[key];
    }
  });
});

var _right = /*#__PURE__*/require("./right.js");

Object.keys(_right).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _right[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _right[key];
    }
  });
});

var _runtime = /*#__PURE__*/require("./runtime.js");

Object.keys(_runtime).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _runtime[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _runtime[key];
    }
  });
});

var _sandbox = /*#__PURE__*/require("./sandbox.js");

Object.keys(_sandbox).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sandbox[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sandbox[key];
    }
  });
});

var _sandboxWith = /*#__PURE__*/require("./sandboxWith.js");

Object.keys(_sandboxWith).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sandboxWith[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sandboxWith[key];
    }
  });
});

var _scopeWith = /*#__PURE__*/require("./scopeWith.js");

Object.keys(_scopeWith).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _scopeWith[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _scopeWith[key];
    }
  });
});

var _second = /*#__PURE__*/require("./second.js");

Object.keys(_second).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _second[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _second[key];
    }
  });
});

var _sleep = /*#__PURE__*/require("./sleep.js");

Object.keys(_sleep).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sleep[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sleep[key];
    }
  });
});

var _some = /*#__PURE__*/require("./some.js");

Object.keys(_some).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _some[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _some[key];
    }
  });
});

var _someOrElse = /*#__PURE__*/require("./someOrElse.js");

Object.keys(_someOrElse).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _someOrElse[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _someOrElse[key];
    }
  });
});

var _someOrElseM = /*#__PURE__*/require("./someOrElseM.js");

Object.keys(_someOrElseM).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _someOrElseM[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _someOrElseM[key];
    }
  });
});

var _someOrFail = /*#__PURE__*/require("./someOrFail.js");

Object.keys(_someOrFail).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _someOrFail[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _someOrFail[key];
    }
  });
});

var _someOrFailException = /*#__PURE__*/require("./someOrFailException.js");

Object.keys(_someOrFailException).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _someOrFailException[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _someOrFailException[key];
    }
  });
});

var _struct = /*#__PURE__*/require("./struct.js");

Object.keys(_struct).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _struct[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _struct[key];
    }
  });
});

var _summarized = /*#__PURE__*/require("./summarized.js");

Object.keys(_summarized).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _summarized[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _summarized[key];
    }
  });
});

var _swap = /*#__PURE__*/require("./swap.js");

Object.keys(_swap).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _swap[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _swap[key];
    }
  });
});

var _tap = /*#__PURE__*/require("./tap.js");

Object.keys(_tap).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tap[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tap[key];
    }
  });
});

var _tapBoth = /*#__PURE__*/require("./tapBoth.js");

Object.keys(_tapBoth).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tapBoth[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tapBoth[key];
    }
  });
});

var _tapCause = /*#__PURE__*/require("./tapCause.js");

Object.keys(_tapCause).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tapCause[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tapCause[key];
    }
  });
});

var _tapEither = /*#__PURE__*/require("./tapEither.js");

Object.keys(_tapEither).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tapEither[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tapEither[key];
    }
  });
});

var _tapError = /*#__PURE__*/require("./tapError.js");

Object.keys(_tapError).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tapError[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tapError[key];
    }
  });
});

var _tapExit = /*#__PURE__*/require("./tapExit.js");

Object.keys(_tapExit).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tapExit[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tapExit[key];
    }
  });
});

var _timed = /*#__PURE__*/require("./timed.js");

Object.keys(_timed).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _timed[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _timed[key];
    }
  });
});

var _timeout = /*#__PURE__*/require("./timeout.js");

Object.keys(_timeout).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _timeout[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _timeout[key];
    }
  });
});

var _timeoutFail = /*#__PURE__*/require("./timeoutFail.js");

Object.keys(_timeoutFail).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _timeoutFail[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _timeoutFail[key];
    }
  });
});

var _timeoutFailCause = /*#__PURE__*/require("./timeoutFailCause.js");

Object.keys(_timeoutFailCause).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _timeoutFailCause[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _timeoutFailCause[key];
    }
  });
});

var _timeoutTo = /*#__PURE__*/require("./timeoutTo.js");

Object.keys(_timeoutTo).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _timeoutTo[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _timeoutTo[key];
    }
  });
});

var _to = /*#__PURE__*/require("./to.js");

Object.keys(_to).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _to[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _to[key];
    }
  });
});

var _toLayer = /*#__PURE__*/require("./toLayer.js");

Object.keys(_toLayer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toLayer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _toLayer[key];
    }
  });
});

var _toLeft = /*#__PURE__*/require("./toLeft.js");

Object.keys(_toLeft).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toLeft[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _toLeft[key];
    }
  });
});

var _toManaged = /*#__PURE__*/require("./toManaged.js");

Object.keys(_toManaged).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toManaged[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _toManaged[key];
    }
  });
});

var _traceMask = /*#__PURE__*/require("./traceMask.js");

Object.keys(_traceMask).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _traceMask[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _traceMask[key];
    }
  });
});

var _tryCatchOption = /*#__PURE__*/require("./tryCatchOption.js");

Object.keys(_tryCatchOption).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tryCatchOption[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tryCatchOption[key];
    }
  });
});

var _tuple = /*#__PURE__*/require("./tuple.js");

Object.keys(_tuple).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tuple[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tuple[key];
    }
  });
});

var _uncause = /*#__PURE__*/require("./uncause.js");

Object.keys(_uncause).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _uncause[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _uncause[key];
    }
  });
});

var _union = /*#__PURE__*/require("./union.js");

Object.keys(_union).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _union[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _union[key];
    }
  });
});

var _unless = /*#__PURE__*/require("./unless.js");

Object.keys(_unless).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _unless[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _unless[key];
    }
  });
});

var _unoption = /*#__PURE__*/require("./unoption.js");

Object.keys(_unoption).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _unoption[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _unoption[key];
    }
  });
});

var _unrefine = /*#__PURE__*/require("./unrefine.js");

Object.keys(_unrefine).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _unrefine[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _unrefine[key];
    }
  });
});

var _unsandbox = /*#__PURE__*/require("./unsandbox.js");

Object.keys(_unsandbox).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _unsandbox[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _unsandbox[key];
    }
  });
});

var _unsome = /*#__PURE__*/require("./unsome.js");

Object.keys(_unsome).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _unsome[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _unsome[key];
    }
  });
});

var _updateService = /*#__PURE__*/require("./updateService.js");

Object.keys(_updateService).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _updateService[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _updateService[key];
    }
  });
});

var _validate = /*#__PURE__*/require("./validate.js");

Object.keys(_validate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _validate[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _validate[key];
    }
  });
});

var _validateFirst = /*#__PURE__*/require("./validateFirst.js");

Object.keys(_validateFirst).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _validateFirst[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _validateFirst[key];
    }
  });
});

var _when = /*#__PURE__*/require("./when.js");

Object.keys(_when).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _when[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _when[key];
    }
  });
});

var _whenCase = /*#__PURE__*/require("./whenCase.js");

Object.keys(_whenCase).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _whenCase[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _whenCase[key];
    }
  });
});

var _whenM = /*#__PURE__*/require("./whenM.js");

Object.keys(_whenM).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _whenM[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _whenM[key];
    }
  });
});

var _withChildren = /*#__PURE__*/require("./withChildren.js");

Object.keys(_withChildren).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _withChildren[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _withChildren[key];
    }
  });
});

var _zip = /*#__PURE__*/require("./zip.js");

Object.keys(_zip).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _zip[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _zip[key];
    }
  });
});

var _zipPar = /*#__PURE__*/require("./zipPar.js");

Object.keys(_zipPar).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _zipPar[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _zipPar[key];
    }
  });
});

var _zips = /*#__PURE__*/require("./zips.js");

Object.keys(_zips).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _zips[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _zips[key];
    }
  });
});

var _zipWith = /*#__PURE__*/require("./zipWith.js");

Object.keys(_zipWith).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _zipWith[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _zipWith[key];
    }
  });
});

var _zipWithPar = /*#__PURE__*/require("./zipWithPar.js");

Object.keys(_zipWithPar).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _zipWithPar[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _zipWithPar[key];
    }
  });
});

var _exclForEach = /*#__PURE__*/require("./excl-forEach.js");
