// @effect-ts/system/RefM/effect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "as", {
  enumerable: true,
  get: function () {
    return _as.as;
  }
});
Object.defineProperty(exports, "asSomeError", {
  enumerable: true,
  get: function () {
    return _asSomeError.asSomeError;
  }
});
Object.defineProperty(exports, "as_", {
  enumerable: true,
  get: function () {
    return _as.as_;
  }
});
Object.defineProperty(exports, "bind", {
  enumerable: true,
  get: function () {
    return _do.bind;
  }
});
Object.defineProperty(exports, "bind_", {
  enumerable: true,
  get: function () {
    return _do.bind_;
  }
});
Object.defineProperty(exports, "chain", {
  enumerable: true,
  get: function () {
    return _core.chain;
  }
});
Object.defineProperty(exports, "chain_", {
  enumerable: true,
  get: function () {
    return _core.chain_;
  }
});
Object.defineProperty(exports, "do", {
  enumerable: true,
  get: function () {
    return _do.do;
  }
});
Object.defineProperty(exports, "fail", {
  enumerable: true,
  get: function () {
    return _fail.fail;
  }
});
Object.defineProperty(exports, "foldM", {
  enumerable: true,
  get: function () {
    return _foldM.foldM;
  }
});
Object.defineProperty(exports, "foldM_", {
  enumerable: true,
  get: function () {
    return _foldM.foldM_;
  }
});
Object.defineProperty(exports, "fromEither", {
  enumerable: true,
  get: function () {
    return _fromEither.fromEither;
  }
});
Object.defineProperty(exports, "ifM_", {
  enumerable: true,
  get: function () {
    return _ifM.ifM_;
  }
});
Object.defineProperty(exports, "map", {
  enumerable: true,
  get: function () {
    return _map.map;
  }
});
Object.defineProperty(exports, "mapError_", {
  enumerable: true,
  get: function () {
    return _mapError.mapError_;
  }
});
Object.defineProperty(exports, "map_", {
  enumerable: true,
  get: function () {
    return _map.map_;
  }
});
Object.defineProperty(exports, "succeed", {
  enumerable: true,
  get: function () {
    return _core.succeed;
  }
});
Object.defineProperty(exports, "toManaged", {
  enumerable: true,
  get: function () {
    return _toManaged.toManaged;
  }
});

var _as = /*#__PURE__*/require("../Effect/as.js");

var _asSomeError = /*#__PURE__*/require("../Effect/asSomeError.js");

var _core = /*#__PURE__*/require("../Effect/core.js");

var _do = /*#__PURE__*/require("../Effect/do.js");

var _fail = /*#__PURE__*/require("../Effect/fail.js");

var _foldM = /*#__PURE__*/require("../Effect/foldM.js");

var _fromEither = /*#__PURE__*/require("../Effect/fromEither.js");

var _ifM = /*#__PURE__*/require("../Effect/ifM.js");

var _map = /*#__PURE__*/require("../Effect/map.js");

var _mapError = /*#__PURE__*/require("../Effect/mapError.js");

var _toManaged = /*#__PURE__*/require("../Effect/toManaged.js");
